File: sapphire/examples/kvstore/src/main/js/sapphire/demo/KeyValueStoreClient.js
/**
 * Client class for testing {@link KeyValueStore}
 */

function main() {
    system = Java.type("java.lang.System")
    var jsHome = system.getenv("JS_HOME")
    var kernelServerIP = system.getenv("KERNEL_SERVER_IP")
    var kernelServerPort = system.getenv("KERNEL_SERVER_PORT")
    var omsIP = system.getenv("OMS_IP")
    var omsPort = system.getenv("OMS_PORT")

    system.setProperty("JS_HOME", jsHome)
    var fileName = jsHome + "/KeyValueStore.js"

    var ClientStub = Java.type("sapphire.demo.stubs.Client_Stub")
    var clientStub = new ClientStub()
    var kvs = clientStub.GetKeyValueStore(kernelServerIP, kernelServerPort, omsIP, omsPort, "js", fileName);

    for (i=0; i<30; ++i) {
        key = "key_" + i;
        val = "val_" + i;

        console.log(`client: setting ${key} = ${val}`)
        kvs.set(key, val)

        var v = kvs.get(key)
        console.log(`client: got value ${v} with key ${key}`)
    }
}

main();

File: sapphire/examples/kvstore/src/main/java/sapphire/demo/KeyValueStoreClient.java
package sapphire.demo;

import sapphire.common.SapphireObjectID;
import sapphire.kernel.server.KernelServerImpl;
import sapphire.oms.OMSServer;

import java.io.File;
import java.net.InetSocketAddress;
import java.nio.file.Files;
import java.rmi.registry.LocateRegistry;
import java.rmi.registry.Registry;
import java.util.List;

/**
 * Client class for testing {@link KeyValueStore}
 */
public class KeyValueStoreClient {
    public static void main(String[] args) throws Exception {
        OMSServer oms = getOMS(args[0], args[1]);

        SapphireObjectID oid = oms.createSapphireObject(getSpec());
        KeyValueStore store = (KeyValueStore)oms.acquireSapphireObjectStub(oid);

        for (int i=0; i<30; ++i) {
            String key = "key_" + i;
            String val = "val_" + i;

            System.out.println(String.format("<Client> setting %s = %s", key, val));
            store.set(key, val);
            val = String.valueOf(store.get(key));
            System.out.println(String.format("<Client> got value %s with key %s", val, key));
        }
    }

    private static OMSServer getOMS(String omsIp, String omsPort) throws Exception {
        new KernelServerImpl(new InetSocketAddress("127.0.0.2", 11111), new InetSocketAddress(omsIp, Integer.parseInt(omsPort)));
        Registry registry = LocateRegistry.getRegistry(omsIp, Integer.parseInt(omsPort));
        OMSServer omsserver = (OMSServer) registry.lookup("SapphireOMS");
        return omsserver;
    }

    private static String getSpec() throws Exception {
        ClassLoader classLoader = new KeyValueStoreClient().getClass().getClassLoader();
        File file = new File(classLoader.getResource("KeyValueStore.yaml").getFile());
        List<String> lines = Files.readAllLines(file.toPath());
        return String.join("\n", lines);
    }
}
